### ISSUE TITLE
Puma won't die when killing rails server with ctrl-c

### ISSUE THREAD

**Issue by AndrewRayCode** [puma/puma#1247]
### Steps to reproduce

1) run a tiny rails server, with puma configured in the Procfile `web: bundle exec puma -C config/puma.rb`

2) start the rails server

3) kill the server in the terminal with ctrl-c

### Expected behavior

The server should shut down

### Actual behavior

Puma hangs with `Gracefully shutting down workers...`. It's undead, can't be killed with another ctrl-c nor a ctrl-d. Even manually finding the processes by pid and killing them won't unfreeze the server. Somehow puma can't even be killed with `pkill -f puma`

### System configuration

**Ruby version**:
Ruby ruby 2.3.3

**Rails version**:
Rails 5.0.1

**Comment by nateberkopec** [comment]
Cannot reproduce. 99% certain this is an issue on your end, rather than Puma. What happens if you send SIGKILL? `pkill -9 -f puma`

**Comment by nateberkopec** [comment]
Also post output of `ps aux | grep puma` after you've tried to terminate and it doesn't work.

**Comment by AndrewRayCode** [comment]
Ah, I didn't know you can pass -9 to pkill. That finally gets rid of it.

Also, lots of rb-fsevent processes stay open too. I kill everything with:

    pkill -9 -f 'rb-fsevent|rails|spring|puma'

Output when it's stuck in shutdown mode:

    $ ps aux | grep puma
             3053  98.8  1.0  2757276 161812 s000  R+    9:24PM   0:17.78 puma: cluster worker 1: 3027 [rails-app]
             3027   0.0  0.5  2560964  77584 s000  S+    9:24PM   0:02.46 puma 3.7.0 (tcp://0.0.0.0:3000) [rails-app]
             3363   0.0  0.0  2451236   2064 s006  S+    9:30PM   0:00.00 grep puma

**Comment by nateberkopec** [comment]
I'm going to need more information or a reproducible container to help you, as I cannot reproduce a failure to respond to SIGTERM. My guess if you've got something else in your app that's causing this behavior.

**Comment by adisos** [comment]
Maybe this is related...

https://github.com/puma/puma/pull/1252

Look at this please, @nateberkopec

**Comment by viamin** [comment]
I'm also seeing puma processes that won't quit. Also, restarting rails servers will cause a worker to spin up the CPUs to 100%. I only started seeing this after upgrading my laptop. 

I've attached a couple of samples from the two cases, which have a couple of similar threads. 
[puma_sample_100CPU.txt](https://github.com/puma/puma/files/1105937/puma_sample_100CPU.txt)
[puma_sample_wont_quit.txt](https://github.com/puma/puma/files/1105938/puma_sample_wont_quit.txt)

**Comment by jvanbaarsen** [comment]
I'm noticing the same problem, but only if I have workers enabled in my config.

This is my config: https://gist.github.com/fe98e60c5cceff865cd0a9f485bda599

When I comment line 8 it works. Hope this helps a little bit

**Comment by ericboehs** [comment]
I believe this is the same issue as https://github.com/puma/puma/issues/1046.

**Comment by jameskirsop** [comment]
I'm also experiencing this when using [Oxidized](https://github.com/ytti/oxidized) from the CLI.

I end up having to open up another terminal session to issue a kill command to quit the instance.

**Comment by Machinerium** [comment]
I'm also experiencing this when using Open3 on a routine (ping command)

**Comment by tadman** [comment]
I've encountered this problem as well and it looks like there's really no easy way to stop Puma from patching in with `Signal.trap`. Having an option to prevent it from doing this would make life a lot more pleasant than having to go to great lengths to prevent the `SIGTERM` trap from being injected.

In development mode graceful shutdown is more trouble than it's worth, so it'd be great if that feature could be disabled via options.

**Comment by ericboehs** [comment]
I've found sending SIGQUIT (via `Ctrl-\`) will kill the process.

**Comment by gconwayoxblue** [comment]
Can confirm SIGQUIT worked for me.

**Comment by tadman** [comment]
The only way to avoid Puma intercepting `SIGINT` and trying to be helpful is to patch out the `Signal.trap` method entirely. It's messy and should be unnecessary, but if anyone needs to keep Puma on a leash, this is how you can do it:

```
module Signal
  def self.trap(signal_name)
  end
end
```

Declaring that before `require "puma"` will prevent it from hooking in. Note that this may interfere with other things that need to `trap` but thankfully those situations are rare.

**Comment by eaconway** [comment]
Not sure if there's a sure-fire way to solve this or if the solution is simply to send SIGQUIT (via `Ctrl-\`), but I'm still having this issue as well. 

Does anyone know why these workers stop synching, specifically in the case that was mentioned 8 comments up on this [gist](https://gist.github.com/fe98e60c5cceff865cd0a9f485bda599), where you have set "Web_concurrrency" to { 2 }?

**Comment by jmrepetti** [comment]
I had the same issue with Puma 3.10.0, updating puma fixed the problem for me.

**Comment by Maclenn77** [comment]
I often have this issue. If I type 

`pkill -9 -f puma `

it solves, but it's annoying that I don't know when CTRL + C will work or not.

**Comment by jesselawson** [comment]
Well it's 2022 and the only pattern I can find with this issue is this: if I run `rails server` from inside the VSCode terminal, a `Ctrl+C` kills the server _but_ `puma` continues to run. The only thing that kills it in this case is a `pkill -9 -f puma`. 

If, however, I run `rails server` in a separate `zsh` window, entering `Ctrl+C` will kill puma as expected.

**Comment by diadras** [comment]
I am using AWS beanstalk to host an application that has 0 relation to Puma or a Ruby Rails webserver and my issue was my machine having a Puma process running that was using 100% CPU. Searching for an unkillable Puma process resulted in this Github thread and little other results.

Killing the process with `pkill -9 -f puma` would make it spin up again and momentarily use even more CPU resources (200%).
I figured out after killing it right when it spun up again too 200% that it is probably being used by the healthd service from Amazon itself. The location I saw scrolling past my screen was `/opt/elasticbeanstalk/lib/ruby/bin/healthd/`

Using `pkill -9 -f puma` a few times together with checking if it spun up again on a second window eventually worked for some reason. The process eventually stopped using 100%/200% CPU and is now running neatly idle.

Sorry if this is completely unrelated but I hope this will be of use for someone having the same issue I had :)
